MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c

```diff
@@ -839,103 +1145,213 @@ NvmeControllerInit (
+  DEBUG ((DEBUG_INFO, "    AERL      : 0x%x\n", Private->ControllerData->Aerl));
+  DEBUG ((DEBUG_INFO, "    SQES      : 0x%x\n", Private->ControllerData->Sqes));
+  DEBUG ((DEBUG_INFO, "    CQES      : 0x%x\n", Private->ControllerData->Cqes));
+  DEBUG ((DEBUG_INFO, "    NN        : 0x%x\n", Private->ControllerData->Nn));
 
   //
-  // Program admin queue attributes.
+  // Send Set Features Command to request the maximum number of data queues.
+  // The controller is free to allocate a different number of queues from the number requested.
+  // The number of queues allocated is returned and stored in the controller private data structure
+  // using the Nsqa and Ncqa fields.
   //
-  Status = WriteNvmeAdminQueueAttributes (Private, &Aqa);
+  Status = NvmeSetFeaturesNumberOfQueues (Private, NVME_MAX_QUEUES - 1, NVME_MAX_QUEUES - 1);
 
   if (EFI_ERROR (Status)) {
     return Status;
   }
 
+  // MU_CHANGE [BEGIN] - Allocate IO Queue Buffer
   //
-  // Program admin submission queue address.
+  // Allocate Data Queues - note we are assuming the queue entry sizes are the same as the admin queue entry sizes for the sake of memory allocation.
+  // The identify controller data tells us in SQES and CQES what the controller's minimum and maximum queue entry sizes are. We haven't used this before since we
+  // use the spec-defined minimum queue entry sizes.
+  // We are also allocating based on the admin defined queue sizes in number of entries.
+  // Some scenarios may use different queue sizes, currently we only see the case where the driver needs IO queue sizes <= admin queue sizes. So this allocation should be sufficient.
+  // We may want to explore a more dynamic allocation in the future.
   //
-  Status = WriteNvmeAdminSubmissionQueueBaseAddress (Private, &Asq);
+  for (Index = 1; Index <= Private->NumberOfDataQueuePairs; Index++) {
+    Private->SqData[Index].NumberOfEntries = Private->SqData[0].NumberOfEntries;
+    Private->CqData[Index].NumberOfEntries = Private->CqData[0].NumberOfEntries;
+    Private->SqData[Index].EntrySize       = Private->SqData[0].EntrySize;
+    Private->CqData[Index].EntrySize       = Private->CqData[0].EntrySize;
+  }
+
+  // Using the first data queue size for the number of pages required for the data queues
+  QueuePairPageCount = EFI_SIZE_TO_PAGES (Private->SqData[1].NumberOfEntries * 2^Private->SqData[1].EntrySize)
+                       + EFI_SIZE_TO_PAGES (Private->CqData[1].NumberOfEntries * 2^Private->CqData[1].EntrySize);
 
+  Status = PciIo->AllocateBuffer (
+                    PciIo,
+                    AllocateAnyPages,
+                    EfiBootServicesData,
+                    QueuePairPageCount * Private->NumberOfDataQueuePairs,
+                    (VOID **)&Private->DataQueueBuffer,
+                    0
+                    );
   if (EFI_ERROR (Status)) {
     return Status;
   }
 
+  DEBUG ((DEBUG_INFO, "Private->DataQueueBuffer = [%016X]\n", (UINT64)(UINTN)Private->DataQueueBuffer));

```

Comment:
Probably don't need this print checked in

Reply:
Copied from https://github.com/microsoft/mu_basecore/blob/dev/202405/MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c#L901
thinking it might be helpful info. Will defer to your judgement though.

Reply:
I don't think that provides any value in the log, it will just print the address of the buffer, which from a log doesn't tell you anything (other prints, queue size, whether BlockIo2 was installed or not, etc. do provide info from the log point of view).